pkg/ansible/runner/runner.go

```diff
@@ -0,0 +1,290 @@
+		Parameters: r.makeParameters(u),
+		EnvVars: map[string]string{
+			"K8S_AUTH_KUBECONFIG": kubeconfig,
+		},
+		Settings: map[string]string{
+			"runner_http_url":  receiver.SocketPath,
+			"runner_http_path": receiver.URLPath,
+		},
+	}
+	// If Path is a dir, assume it is a role path. Otherwise assume it's a
+	// playbook path
+	fi, err := os.Lstat(r.Path)
+	if err != nil {
+		return nil, err
+	}
+	if !fi.IsDir() {
+		inputDir.PlaybookPath = r.Path
+	}
+	err = inputDir.Write()
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		var dc *exec.Cmd
+		if r.isFinalizerRun(u) {
+			logger.Debugf("Resource is marked for deletion, running finalizer %s", r.Finalizer.Name)
+			dc = r.finalizerCmdFunc(ident, inputDir.Path)
+		} else {
+			dc = r.cmdFunc(ident, inputDir.Path)
+		}
+
+		err := dc.Run()
+		if err != nil {
+			logger.Errorf("error from ansible-runner: %s", err.Error())
+		} else {
+			logger.Info("ansible-runner exited successfully")
+		}
+
+		receiver.Close()
+		err = <-errChan
+		// http.Server returns this in the case of being closed cleanly
+		if err != nil && err != http.ErrServerClosed {
+			logger.Errorf("error from event api: %s", err.Error())
+		}
+	}()
+	return receiver.Events, nil
+}
+
+func (r *runner) GetFinalizer() (string, bool) {
+	if r.Finalizer != nil {
+		return r.Finalizer.Name, true
+	}
+	return "", false
+}
+
+func (r *runner) isFinalizerRun(u *unstructured.Unstructured) bool {
+	finalizersSet := r.Finalizer != nil && u.GetFinalizers() != nil
+	// The the resource is deleted and our finalizer is present, we need to run the finalizer
+	if finalizersSet && u.GetDeletionTimestamp() != nil {
+		for _, f := range u.GetFinalizers() {
+			if f == r.Finalizer.Name {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (r *runner) addFinalizer(finalizer *Finalizer) error {
+	r.Finalizer = finalizer
+	switch {
+	case finalizer == nil:
+		return nil
+	case finalizer.Playbook != "":
+		if !filepath.IsAbs(finalizer.Playbook) {
+			return fmt.Errorf("finalizer playbook path must be absolute for %v", r.GVK)
+		}
+		r.finalizerCmdFunc = func(ident, inputDirPath string) *exec.Cmd {
+			return exec.Command("ansible-runner", "-vv", "-p", finalizer.Playbook, "-i", ident, "run", inputDirPath)
+		}
+	case finalizer.Role != "":
+		if !filepath.IsAbs(finalizer.Role) {
+			return fmt.Errorf("finalizer role path must be absolute for %v", r.GVK)
+		}
+		r.finalizerCmdFunc = func(ident, inputDirPath string) *exec.Cmd {
+			path := strings.TrimRight(finalizer.Role, "/")
+			rolePath, roleName := filepath.Split(path)
+			return exec.Command("ansible-runner", "-vv", "--role", roleName, "--roles-path", rolePath, "--hosts", "localhost", "-i", ident, "run", inputDirPath)
+		}
+	case len(finalizer.Vars) != 0:
+		r.finalizerCmdFunc = r.cmdFunc
+	}
+	return nil
+}
+func (r *runner) makeParameters(u *unstructured.Unstructured) map[string]interface{} {
```

Comment:
It might be helpful to have a comment here to explain what the output format for parameters is:
```
{ "meta": {
        "name": "<cr-name>",
        "namespace": "<cr-namespace>",
  },
 <CR.spec fields>,
  "_<group>_<kind>": {
     <Full CR contents>
   },
}
```
At least that's what I think it ends up as in the input directory's env/extravars.